File: studio/screens/history.js
// Phase 5 — History: scrub a recorded journal (GET /playback.json, only
// present when the server was started with --playback demo.db). Keeps its
// own local reconstructed state rather than touching the live store
// singleton, so scrubbing the past never corrupts what Home/Water/Energy
// show as "now" if the user switches tabs mid-scrub.
import { h, mount } from "../core/dom.js";
import { fmt, fmtTick, STATE } from "../core/format.js";
import { chip, healthDot } from "../ui/health-dot.js";

const PLAY_INTERVAL_MS = 200;

let target = null;
let timeline = [];
let tMax = 0;
let t = 0;
let playing = false;
let timer = null;

export function renderHistory(mountTarget) {
  target = mountTarget;
  stop();
  timeline = []; tMax = 0; t = 0;
  paint("loading");

  fetch("/playback.json")
    .then((res) => (res.ok ? res.json() : Promise.reject()))
    .then((data) => {
      timeline = data.timeline || [];
      tMax = data.t_max || 0;
      t = tMax; // land on the resolved end-state; scrub back to watch it unfold
      paint();
    })
    .catch(() => paint("unavailable"));
}

function paint(mode) {
  if (!target) return;
  mount(target, view(mode));
}

function reconstructAt(tick) {
  const points = new Map();
  const actions = new Map();
  const actionOrder = [];
  for (const f of timeline) {
    if (f.t > tick) break; // JournalReader.timeline() is sorted by t
    if (f.type === "point") {
      points.set(f.key, { value: f.value, t: f.t, source: f.source, confidence: f.confidence });
    } else if (f.type === "event") {
      const payload = f.payload || {};
      const a = payload.action;
      if ((f.topic === "domora/plan/action" || f.topic === "domora/act/dispatched") && a) {
        let m = actions.get(a.id);
        if (!m) { m = { id: a.id, status: "pending" }; actions.set(a.id, m); actionOrder.unshift(a.id); }
        m.cause = a.cause; m.command = a.command_topic; m.evidence = a.evidence;
        if (a.expectation) m.expect = a.expectation.describe;
      } else if (f.topic === "domora/verify/confirmed" && actions.has(payload.action_id)) {
        actions.get(payload.action_id).status = "confirmed";
      } else if (f.topic.startsWith("domora/alert") && actions.has(payload.action_id)) {
        Object.assign(actions.get(payload.action_id), { status: "failed", reason: payload.reason });
      }
    }
  }
  return { points, actions, actionOrder };
}

function seek(tick) {
  t = Math.max(0, Math.min(tMax, tick));
  paint();
}

function stop() {
  playing = false;
  if (timer) clearInterval(timer);
  timer = null;
}

function play() {
  if (t >= tMax) seek(0);
  playing = true;
  timer = setInterval(() => {
    if (t >= tMax) { stop(); paint(); return; }
    seek(t + 1);
  }, PLAY_INTERVAL_MS);
  paint();
}

function view(mode) {
  if (mode === "loading") {
    return h("div", { class: "glass card" }, [h("h2", {}, "History"), h("div", { class: "empty" }, "Loading recorded journal…")]);
  }
  if (mode === "unavailable" || !timeline.length) {
    return h("div", { class: "glass card" }, [
      h("h2", {}, "History"),
      h("div", { class: "empty" },
        "No recorded journal for this run. Home, Water, Energy, and Security already show what's happening now. " +
        "To browse history: run the hub with --journal demo.db, then serve it with " +
        "`python -m hub.services.api --playback demo.db`."),
    ]);
  }

  const { points, actions, actionOrder } = reconstructAt(t);
  const keys = [...points.keys()].sort();

  return h("div", { class: "history-screen" }, [
    h("div", { class: "glass card history-scrubber" }, [
      h("div", { class: "history-controls" }, [
        h("button", { type: "button", class: "history-play", onclick: () => (playing ? stop() : play()) }, playing ? "❚❚ Pause" : "▶ Play"),
        h("input", {
          type: "range", min: 0, max: tMax, value: t, class: "history-range",
          oninput: (e) => { stop(); seek(Number(e.target.value)); },
        }),
        h("span", { class: "history-t" }, `${fmtTick(t)} / ${fmtTick(tMax)}`),
      ]),
    ]),

    h("div", { class: "glass card" }, [
      h("h2", {}, "Actions at this point"),
      actionOrder.length
        ? h("div", { class: "incident-list" }, actionOrder.map((id) => actionRow(actions.get(id))))
        : h("div", { class: "empty" }, "No action dispatched yet at this tick."),
    ]),

    h("div", { class: "glass card" }, [
      h("h2", {}, "Points at this point"),
      h("table", { class: "sensor-table" }, [
        h("thead", {}, [h("tr", {}, [h("th", {}, "key"), h("th", {}, "value")])]),
        h("tbody", {}, keys.map((k) => h("tr", {}, [h("td", {}, k), h("td", {}, fmt(points.get(k).value))]))),
      ]),
    ]),
  ]);
}

function actionRow(a) {
  const statusChip = a.status === "confirmed" ? chip("verified", STATE.OK)
    : a.status === "failed" ? chip("escalated", STATE.CRIT)
    : chip("pending", STATE.LEARN);
  return h("div", { class: "incident-row" }, [
    h("div", { class: "incident-head" }, [
      healthDot(a.status === "confirmed" ? STATE.OK : a.status === "failed" ? STATE.CRIT : STATE.LEARN),
      h("span", { class: "incident-cause" }, a.cause || `action #${a.id}`),
      statusChip,
    ]),
  ]);
}

File: mobile/lib/screens/history.dart
// Phase E — History: scrub a recorded journal (GET /playback.json, present
// only when the hub was started with --playback demo.db). Port of
// studio/screens/history.js. Keeps its own reconstructed state rather than
// touching the live store, so scrubbing the past never corrupts what
// Home/Water/Energy show as "now" when the user switches tabs mid-scrub.
import 'dart:async';

import 'package:flutter/material.dart';

import '../core/app_scope.dart';
import '../core/format.dart';
import '../core/playback.dart';
import '../theme/tokens.dart';
import '../widgets/glass_card.dart';
import '../widgets/incident_row.dart';

const _playInterval = Duration(milliseconds: 200);

class HistoryScreen extends StatefulWidget {
  const HistoryScreen({super.key});

  @override
  State<HistoryScreen> createState() => _HistoryScreenState();
}

class _HistoryScreenState extends State<HistoryScreen> {
  PlaybackTimeline? _timeline;
  bool _loaded = false;
  int _t = 0;
  bool _playing = false;
  Timer? _timer;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) => _load());
  }

  Future<void> _load() async {
    final timeline = PlaybackTimeline.parse(await HubScope.of(context).fetchPlayback());
    if (!mounted) return;
    setState(() {
      _timeline = timeline;
      _loaded = true;
      // Land on the resolved end-state; scrub back to watch it unfold.
      _t = timeline?.tMax ?? 0;
    });
  }

  void _stop() {
    _timer?.cancel();
    _timer = null;
    if (_playing) setState(() => _playing = false);
  }

  void _play() {
    final tMax = _timeline?.tMax ?? 0;
    if (_t >= tMax) setState(() => _t = 0);
    setState(() => _playing = true);
    _timer = Timer.periodic(_playInterval, (_) {
      if (_t >= tMax) {
        _stop();
        return;
      }
      setState(() => _t += 1);
    });
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('History')),
      body: _body(),
    );
  }

  Widget _body() {
    final timeline = _timeline;
    if (!_loaded) {
      return const Padding(
        padding: EdgeInsets.all(DomoraSpace.s4),
        child: GlassCard(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [CardTitle('History'), EmptyNote('Loading recorded journal…')],
          ),
        ),
      );
    }
    if (timeline == null || timeline.isEmpty) {
      return const Padding(
        padding: EdgeInsets.all(DomoraSpace.s4),
        child: GlassCard(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              CardTitle('History'),
              EmptyNote('No recorded journal for this run. Home, Water, Energy, and Security already show what is '
                  'happening now. To browse history: run the hub with --journal demo.db, then serve it with '
                  '`python -m hub.services.api --playback demo.db`.'),
            ],
          ),
        ),
      );
    }

    final state = reconstructAt(timeline, _t);
    final keys = state.points.keys.toList()..sort();

    return ListView(
      padding: const EdgeInsets.all(DomoraSpace.s4),
      children: [
        GlassCard(
          child: Row(
            children: [
              IconButton(
                icon: Icon(_playing ? Icons.pause_rounded : Icons.play_arrow_rounded),
                color: DomoraColors.accent,
                onPressed: () => _playing ? _stop() : _play(),
                tooltip: _playing ? 'Pause' : 'Play',
              ),
              Expanded(
                child: Slider(
                  value: _t.toDouble().clamp(0, timeline.tMax.toDouble()),
                  min: 0,
                  max: timeline.tMax.toDouble(),
                  onChanged: (v) {
                    _stop();
                    setState(() => _t = v.round());
                  },
                ),
              ),
              Text('${fmtTick(_t)} / ${fmtTick(timeline.tMax)}',
                  style: const TextStyle(fontFamily: 'monospace', fontSize: 11, color: DomoraColors.inkFaint)),
            ],
          ),
        ),
        const SizedBox(height: DomoraSpace.s4),
        GlassCard(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const CardTitle('Actions at this point'),
              if (state.actionOrder.isEmpty)
                const EmptyNote('No action dispatched yet at this tick.')
              else
                for (final id in state.actionOrder)
                  IncidentRow(action: state.actions[id]!, detailed: false),
            ],
          ),
        ),
        const SizedBox(height: DomoraSpace.s4),
        GlassCard(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const CardTitle('Points at this point'),
              for (final k in keys)
                Padding(
                  padding: const EdgeInsets.symmetric(vertical: 3),
                  // Some twin points carry evidence maps, whose string form is
                  // far wider than a phone — both columns are bounded and
                  // ellipsized so a long value can't blow up the row.
                  child: Row(
                    children: [
                      Expanded(
                        flex: 3,
                        child: Text(k,
                            overflow: TextOverflow.ellipsis,
                            style: const TextStyle(fontFamily: 'monospace', fontSize: 11, color: DomoraColors.inkDim)),
                      ),
                      const SizedBox(width: DomoraSpace.s2),
                      Expanded(
                        flex: 2,
                        child: Text(fmtValue(state.points[k]!.value),
                            overflow: TextOverflow.ellipsis,
                            textAlign: TextAlign.right,
                            style: const TextStyle(fontFamily: 'monospace', fontSize: 11, color: DomoraColors.ink)),
                      ),
                    ],
                  ),
                ),
            ],
          ),
        ),
      ],
    );
  }
}
